manpages/imodtrans.man
.na
.nh
.TH imodtrans 1 3.6.16 BL3DEMC
.SH NAME
imodtrans \- Transform an IMOD model file.
.SH SYNOPSIS
imodtrans  [options]  input_model  output_model
.SH DESCRIPTION
Applies transformations to an IMOD model and copies the result to a new
IMOD model.  The transformations can be specified in three
different ways: 1) as translations, rotations,
and scaling factors; 2) as a list of 2-D transformations in a file, one for
each section, such as are
produced by Midas(1) and Xfalign(1); 3) as a single 3-D transformation in a
file, such as is produced by Solvematch(1).
When the transformation is specified by individual components, they are
applied in the order that they are entered, and any option can be entered
more than once.
If you want your rotations to match those applied by the 
Slicer in 3dmod(1) or by Rotatevol(1), specify rotations around Z, then Y,
then X.
.P
If you plan to display the transformed model on a different image file from
the one that it was built on, the program will not be able to translate the
model to the correct location unless it knows the dimensions of the other
image file.  If these dimensions differ from those of the original image
file, enter them with the
.B -n
option.  With the dimensions known, the program will translate from the middle
of the original volume to the middle of the new one, and you can specify
additional transformations if that center-to-center mapping is not
appropriate.
.P
The situation may be more complicated when the original and new volumes have
different pixel sizes in their headers.  3dmod(1) takes a change in pixel size
into account when a model is loaded onto a volume, unless coordinate scaling
is overridden with the 
.B -n
option to 3dmod(1).  If you scale a volume in 2D with newstack(1) by binning,
expansion, or squeezing, or in 3D with binvol(1) or squeezevol(1), a model
will display correctly on the scaled volume with no need to run imodtrans on
it.  When it is saved from 3dmod, it will be written with coordinates that
match the new volume.  If you need to transform a model into those
coordinates without having to load it into 3dmod, use imodtrans with the 
.B -i
option and the name of the scaled image file, instead of using the 
.B -sx, -sy, -sz,
and
.B -n 
options. The transformed model will have the scaled coordinates and will also
load correctly in 3dmod.  If for some reason you do use the scaling and new
size options instead, do not use the 
.B -i
option.  The resulting model will have the new coordinates, but it will not load
correctly into 3dmod without the 
.B -m
option.
.P
For 3-D transformations, however they are specified, all contour points are 
transformed, as well as all mesh points, surface normals, and clipping
planes.  The mesh is transformed because formerly it was not possible to
remesh contours that have been rotated out
of the X/Y plane, but this is now possible.  For 2-D transformations, only
contour points are transformed, so clipping planes are not preserved and the
model will have to be remeshed.  However, if a single 2D transformation is 
applied to the whole model it is converted to a 3D transformation and the
entire model is properly transformed.
.P
The program can also manipulate the flipped state of the model in Y and Z 
either by actually flipping the Y and Z coordinates, or by simply toggling the
flag that indicates whether the model was loaded on data that were flipped in
3dmod.  Either way 
may be used after flipping a volume with "clip flipyz" to change a model built 
on the
unflipped volume so that it will load onto the flipped volume.
In addition, the model can be transformed to match a particular image
file, just the same as if it were loaded onto that file in 3dmod.  The order
of operations is: flipping, then transforming to match an image file, then 
transforming either via transformations in a file or via the entered options.
.SH OPTIONS
.TP
.B -tx \fIvalue\fR
.PD 0
.TP
.B -ty \fIvalue\fR
.TP
.B -tz \fIvalue\fR
.PD
The
.B -tx, -ty
and
.B -tz
options can be used to 
translate the model in the particular direction by the given floating point 
value, in pixels.
.TP
.B -sx \fIvalue\fR
.PD 0
.TP
.B -sy \fIvalue\fR
.TP
.B -sz \fIvalue\fR
.PD
The
.B -sx, -sy
and
.B -sz
options can be used to 
scale the model in the particular dimension by the given floating point value.
.TP
.B -rx \fIvalue\fR
.PD 0
.TP
.B -ry \fIvalue\fR
.TP
.B -rz \fIvalue\fR
.PD
The
.B -rx, -ry
and
.B -rz
options can be used to 
rotate the model about the particular axis by the given value in degrees.
.TP
.B -2 \fItransform file\fR
Transform the model in 2-D sections using the given transform file
as input.  One line is expected per section, and slices at Z values greater 
than the
number of lines will not be transformed, unless the 
.B -l
option is given to apply one transform to the whole model.
This option is useful for 
transforming a model to match an
image stack that has been aligned with Xfalign or Midas.
This option cannot be entered with the scaling or rotation
options
or with 
.B -3.
The translation options
.B -tx
and
.B -ty
can be used with this option, and 
.B -tz
can be used when applying one transform to the whole model.  In each case, the
option can be entered only once, and the shift is applied after the transformation.
.TP
.B -l \fIline #\fR
Transform the whole model with the single 2D transformation at the given line
number in the transform file entered with 
.B -2.
Lines are numbered from 0.  

.TP
.B -3 \fItransform file\fR
Transform the model with the single 3-D transformation in the given file.
The format of the file is to have 3 lines, each with 4 numbers, the
coefficients for multiplying the old X, Y, and Z coordinates and a translation
value.  The first, second, and third lines are for computing new X, Y, and
Z coordinates, respectively.
This option cannot be entered with the scaling or rotation
options, or with 
.B -2.
It can be used with the translation options, but each option can be entered
only once, and the shift is applied after the transformation.
.TP
.B -S \fIfactor\fR
Scale the X, Y, and Z shifts in the transformations from transform files by the given
factor.  This option is needed, e.g., if transformations have shifts based on
unbinned coordinates and the model is now in
binned coordinates.
.TP
.B -n \fINX,NY,NZ\fR
Specify the dimensions of the image file on which the model is to be
displayed, if these dimensions are not the same as for the image file
on which the model was originally built.  This entry is needed to
have the model translated to the correct location when image file sizes
differ.
.TP
.B -z
Apply Z-scaling to the model, using the Z-scale from the model header, before
transforming the model.  The transformed model will then be descaled in Z so
that it will display correctly with the same Z-scale.  This option would be
needed to retain the shape of a Z-scaled model, but it would probably prevent
it from matching a transformed image file.
.TP
.B -f
If the model was last displayed on images loaded into 3dmod with Y and Z
flipped, use this option to apply the transformation to the flipped
coordinates.  The default is to apply the transformation to native, unflipped
coordinates, which will work with transformations obtained from Solvematch.
.TP
.B -i \fIimage file\fR
Transform the model coordinates as if the model were being loaded into 3dmod
together with the given image file.  The maximum coordinates in the model
(reported as "max # # #" in the output from Imodinfo(1)) will be changed to
match those in the image file.  This transformation occurs before any of
the other operations except flipping with the
.B -Y 
option.
.TP
.B -Y
Exchange Y and Z coordinates in the model file, without changing the flag
indicating the model's flip state.
.TP
.B -T
Toggle the flag indicating the model's flip state, without actually flipping
the data.

.SH Examples
Example 1: Suppose a model is drawn on a raw stack (set.st) and needs to
be transformed to fit on a binned by 2 stack (set.ali) that was aligned with
transformations in set.xg.  The aligned file might be a different size
in X and Y than the size of the raw stack divided by the binning.  If the
origin was adjusted when the aligned stack was created, using the 
.B -origin
option to Newstack(1) or Blendmont(1), then the command would be:
   imodtrans -2 set.xg -i set.ali -S 0.5 input_model output_model
.br
First, the program reduces all X and Y coordinates by 2 to match the
coordinate system in set.ali, and shifts points based on the change in origin
values.  Then it multiples the shifts in set.xg by 0.5, since they are in
unbinned coordinates, and applies the transformations.

However, if the origin was not adjusted for any size change, then it is
necessary to add a translation to account for it.  The needed translation is:
   xtrans = (alignedXsize - rawXsize / binning) / 2.
   ytrans = (alignedYsize - rawYsize / binning) / 2.
.br 
And the command is
   imodtrans -2 set.xg -i set.ali -S 0.5 -tx xtrans -ty ytrans input output

Example 2: Suppose a model is drawn on one stack (seta.ali) and needs to fit 
on another stack (setb.ali) that is rotated by approximately 90 degrees.  The
file set_AtoB.xf has a single line with the transformation that would align
the images.  The command is:
   imodtrans -2 set.xg -l 0 -n bXsize,bYsize,bZsize input_model output_model
.br
This command works because the input model has the size of seta.st stored as
its maximum coordinates, and because the output file size is given.  The
prgram is thus able to shift the model by the difference between the center
coordinates of the two stack.

.SH AUTHOR
Jim Kremer and David Mastronarde
.SH SEE ALSO
midas(1), xfalign(1), solvematch(1), matchvol(1), rotatevol(1), clip(1),
3dmod(1)
.SH BUGS

Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

imodutil/imodtrans.c
/*
 *  imodtrans.c -- Translate, scale and rotate imod model files.
 *
 *  Original Author: James Kremer
 *  Revised by: David Mastronarde   email: mast@colorado.edu
 *
 * $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "imodel.h"
#include "mrcfiles.h"
#include "parse_params.h"


static int filetrans(char *filename, Imod *model, int mode, int oneLine,
                     Ipoint newCen, float zscale, int doflip, float transScale, 
                     float transx, float transy, float transz);
static int trans_model_slice(Imod *model, float *mat, int slice,
                             Ipoint newCen);
static void usage(char *progname);


static void usage(char *progname)
{
  imodVersion(progname);
  imodCopyright();
  printf("Usage: %s [options] <input file> <output file>\n", 
          progname);
  printf("Options:\n");
  printf("\t-tx #\tTranslate in X by #\n");
  printf("\t-ty #\tTranslate in Y by #\n");
  printf("\t-tz #\tTranslate in Z by #\n");
  printf("\t-sx #\tScale in X by #\n");
  printf("\t-sy #\tScale in Y by #\n");
  printf("\t-sz #\tScale in Z by #\n");
  printf("\t-rx #\tRotate around X axis by #\n");
  printf("\t-ry #\tRotate around Y axis by #\n");
  printf("\t-rz #\tRotate around Z axis by #\n");
  printf("\t-2 file\tApply 2D transformations from file, one per"
          " section\n");
  printf("\t-l #\tLine number of single 2D transformation to apply"
          " (from 0)\n");
  printf("\t-3 file\tApply 3D transformation from file\n");
  printf("\t-S #\tScale dx/dy[/dz] values in 2D/3D transformations by #\n");
  printf("\t-n #,#,#\tSet X,Y,Z size of transformed volume\n");
  printf("\t-z\tTransform Z-scaled coordinates using model's Z-scale"
          "\n");
  printf("\t-f\tTransform flipped instead of native coordinates if "
          "Y-Z flipped\n");
  printf("\t-i file\tTransform to match given image file coordinate"
          " system\n");
  printf("\t-Y\tFlip model in Y and Z (without toggling flipped"
          " flag)\n");
  printf("\t-T\tToggle flag that model is flipped in Y and Z\n");

  exit(3);
}

int main(int argc, char *argv[])
{
  Imod   model;
  FILE   *fin, *fout;
  char   *filename = NULL;
  int    i;
  int    mode = 0;
  int    useZscale = 0;
  int    doflip = 1;
  int    transopt = 0;
  int    rotScaleopt = 0;
  int    toImage = 0;
  int    oneLine = -1;
  int    toggleFlip = 0;
  int    flipModel = 0;
  float  zscale = 1.;
  float transScale = 1.;
  float  rx = 0., ry = 0., rz = 0.;
  float transx = 0., transy = 0., transz = 0.;
  int multiTrans = 0;
  int newNx = 0, newNy = 0, newNz = 0;
  char *progname = imodProgName(argv[0]);
  Ipoint newCen, tmpPt;
  Imat *mat = imodMatNew(3);
  Imat *normMat = imodMatNew(3);
  IrefImage useRef, *modRefp;
  MrcHeader hdata;
  Ipoint unitPt = {1., 1., 1.};
  char prefix[100];
  sprintf(prefix, "ERROR: %s - ", progname);
  setExitPrefix(prefix);

  if (argc < 3){
    usage(progname);
  }
     
  for (i = 1; i < argc; i++) {
    if (argv[i][0] == '-') {
      switch (argv[i][1]) {

      case '2':
        filename = argv[++i];
        mode += 2;
        break;
            
      case '3':
        filename = argv[++i];
        mode += 3;
        break;

      case 'S':
        sscanf(argv[++i], "%f", &transScale);
        break;

      case 'i':
        toImage = 1;
        if (NULL == (fin = fopen(argv[++i], "rb")))
          exitError("Couldn't open %s", argv[i]);

        if (mrc_head_read(fin, &hdata)) 
          exitError("Reading header from %s", argv[i]);
        fclose(fin);

      case 'z':
        useZscale = 1;
        break;

      case 'f':
        doflip = 0;
        break;

      case 'Y':
        flipModel = 1;
        break;

      case 'T':
        toggleFlip = 1;
        break;

      case 'n':
        sscanf(argv[++i], "%d%*c%d%*c%d", &newNx, &newNy, &newNz);
        break;
            
      case 'l':
        oneLine = atoi(argv[++i]);
        break;

      case 't':  /* Translations */
        tmpPt.x = tmpPt.y = tmpPt.z = 0.;
        transopt = 1;
        switch (argv[i][2]) {
        case 'x':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-tx%f", &tmpPt.x);
          else
            sscanf(argv[++i], "%f", &tmpPt.x);
          if (transx)
            multiTrans = 1;
          transx = tmpPt.x;
          break;
        case 'y':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-ty%f", &tmpPt.y);
          else
            sscanf(argv[++i], "%f", &tmpPt.y);
          if (transy)
            multiTrans = 1;
          transy = tmpPt.y;
          break;
        case 'z':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-tz%f", &tmpPt.z);
          else
            sscanf(argv[++i], "%f", &tmpPt.z);
          if (transz)
            multiTrans = 1;
          transz = tmpPt.z;
          break;
        default:
          exitError("Invalid option %s", argv[i]);
        }
        imodMatTrans(mat, &tmpPt);
        break;

      case 's':  /* Scaling */
        tmpPt.x = tmpPt.y = tmpPt.z = 1.;
        transopt = 1;
        rotScaleopt = 1;
        switch (argv[i][2]) {
        case 'x':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-sx%f", &tmpPt.x);
          else
            sscanf(argv[++i], "%f", &tmpPt.x);
          break;
        case 'y':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-sy%f", &tmpPt.y);
          else
            sscanf(argv[++i], "%f", &tmpPt.y);
          break;
        case 'z':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-sz%f", &tmpPt.z);
          else
            sscanf(argv[++i], "%f", &tmpPt.z);
          break;
        default:
          exitError("Invalid option %s", argv[i]);
        }
        imodMatScale(mat, &tmpPt);
        tmpPt.x = 1. / tmpPt.x;
        tmpPt.y = 1. / tmpPt.y;
        tmpPt.z = 1. / tmpPt.z;
        imodMatScale(normMat, &tmpPt);
        break;

      case 'r':  /* Rotations */
        transopt = 1;
        rotScaleopt = 1;
        switch (argv[i][2]) {
        case 'x':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-rx%f", &rx);
          else
            sscanf(argv[++i], "%f", &rx);
          imodMatRot(mat, (double)rx, b3dX);
          imodMatRot(normMat, (double)rx, b3dX);
          break;
        case 'y':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-ry%f", &ry);
          else
            sscanf(argv[++i], "%f", &ry);
          imodMatRot(mat, (double)ry, b3dY);
          imodMatRot(normMat, (double)ry, b3dY);
          break;
        case 'z':
          if (argv[i][3] != 0x00)
            sscanf(argv[i], "-rz%f", &rz);
          else
            sscanf(argv[++i], "%f", &rz);
          imodMatRot(mat, (double)rz, b3dZ);
          imodMatRot(normMat, (double)rz, b3dZ);
          break;
        default:
          exitError("Invalid option %s", argv[i]);
        }
        break;

      default:
        exitError("Invalid option %s",  argv[i]);
      }

    }else{
      break;
    }
  }

  if (mode && mode / 2 != 1) 
    exitError("You cannot enter both -2 and -3");
  
  if (mode && rotScaleopt)  
    exitError("You cannot enter -r or -s options with -2 and -3");
  if (transopt && mode == 2 && oneLine < 1 && transz)
    exitError("You cannot enter -tz option with -2 unless transforming with one line");
  if (mode && multiTrans)
    exitError("You cannot enter -t options more than once with -2 and -3");
  
  if (i != argc - 2) {
    printf("ERROR: %s - Command line should end with two non-option "
            "arguments\n", progname);
    usage(progname);
  }

  fin = fopen(argv[i], "rb");
  if (!fin)
    exitError("Opening input file %s", argv[i]);
      
  imodDefault(&model);
  model.file = fin;
  if (imodReadFile(&model))
    exitError("Reading imod model (%s)", argv[i]);
  fclose(fin);

  if (imodBackupFile(argv[i + 1])) 
    exitError("Couldn't create backup file");

  fout = fopen(argv[i + 1], "wb");
  if (!fout)
    exitError("Opening output file %s", argv[i + 1]);

  /* Do flipping operations first */
  if (flipModel)
    imodFlipYZ(&model);
  if (toggleFlip) {
    if (model.flags & IMODF_FLIPYZ)
      model.flags &= ~IMODF_FLIPYZ;
    else
      model.flags |= IMODF_FLIPYZ;
  }

  /* Do scaling to image reference next */
  if (toImage) {

    modRefp = model.refImage;
    if (modRefp && !(model.flags & IMODF_OTRANS_ORIGIN)) 
      exitError("Model has no image origin information; -i option invalid");

    /* get the target transformation */
    useRef.ctrans.x = hdata.xorg;
    useRef.ctrans.y = hdata.yorg;
    useRef.ctrans.z = hdata.zorg;
    useRef.crot.x = hdata.tiltangles[3];
    useRef.crot.y = hdata.tiltangles[4];
    useRef.crot.z = hdata.tiltangles[5];
    useRef.cscale = unitPt;
    if (hdata.xlen && hdata.mx)
      useRef.cscale.x = hdata.xlen/(float)hdata.mx;
    if (hdata.ylen && hdata.my)
      useRef.cscale.y = hdata.ylen/(float)hdata.my;
    if (hdata.zlen && hdata.mz)
      useRef.cscale.z = hdata.zlen/(float)hdata.mz;
    
    if (modRefp) {

      /* If there is a refImage, scale the data as when reading into 3dmod */
      useRef.otrans = modRefp->ctrans;
      useRef.orot = modRefp->crot;
      useRef.oscale = modRefp->cscale;
      imodTransFromRefImage(&model, &useRef, unitPt);
    } else {

      /* If there is no refImage, do not scale data but assign all the information
         just as if reading into 3dmod */
      model.refImage = (IrefImage *)malloc (sizeof(IrefImage));
      modRefp = model.refImage;
      if (!modRefp) 
        exitError("Allocating a IrefImage structure");
      model.flags |= IMODF_OTRANS_ORIGIN;
    }
    *modRefp = useRef;
    modRefp->otrans = useRef.ctrans;

    /* Adjust the maxes for this change */
    model.xmax = hdata.nx;
    model.ymax = hdata.ny;
    model.zmax = hdata.nz;
  }

  /* Use zscale if user indicated it */
  if (useZscale && model.zscale > 0.)
    zscale = model.zscale;

  /* Do flipping if model flipped and user did not turn it off */
  if (!(model.flags & IMODF_FLIPYZ))
    doflip = 0;

  /* Warning every time is too noxious!
  if (!newNx)
    printf("Assuming transformed images have same size as original "
            "images\n (use -n option to specify size of transformed "
            "images)\n");
  */
  newCen.x = (newNx ? newNx : model.xmax) * 0.5f;
  newCen.y = (newNy ? newNy : model.ymax) * 0.5f;
  newCen.z = (newNz ? newNz : model.zmax) * 0.5f;

  if (filename){
    if (filetrans(filename, &model, mode, oneLine, newCen, zscale, doflip, transScale,
                  transx, transy, transz)) 
      exitError("Transforming model.");

  } else if (transopt) {
    imodTransModel3D(&model, mat, normMat, newCen, zscale, doflip);
  }

  model.xmax = newNx ? newNx : model.xmax;
  model.ymax = newNy ? newNy : model.ymax;
  model.zmax = newNz ? newNz : model.zmax;
  imodWrite(&model, fout);
  exit(0);
}

/* 
 * Transform by 2D or 3D transforms from a file
 *  filename  has the name of the file with transforms
 *  model     is the model
 *  mode      is 2 for 2D, 3 for 3D transforms
 *  oneLine   is line number for 2D applied to whole model
 *  newCen    has the new center coordinates of the volume
 *  zscale    is the z scale factor, or 1 not to use any
 *  doflip    indicates transform native form of flipped data
 */
static int filetrans(char *filename, Imod *model, int mode, int oneLine,
                     Ipoint newCen, float zscale, int doflip, float transScale,
                     float transx, float transy, float transz)
{
  FILE *fin;
  char line[80];
  float mat[6];
  int k = 0;
  int nread;
  Imat *mat3d = imodMatNew(3);

  fin = fopen(filename, "r");

  if (!fin){
    printf("ERROR: Couldn't open %s\n", filename);
    return(-1);
  }
     
  if (mode == 2) {
    while (fgetline(fin, line, 80) >= 0){
      
      nread = sscanf(line, "%f %f %f %f %f %f", 
                     &(mat[0]), &(mat[1]), 
                     &(mat[2]), &(mat[3]),
                     &(mat[4]), &(mat[5]));
      if (nread <= 0)
        continue;
      mat[4] = mat[4] * transScale + transx;
      mat[5] = mat[5] * transScale + transy;
      if (oneLine < 0) {
        trans_model_slice(model, mat, k, newCen);
      } else if (k == oneLine) {
        mat3d->data[0] = mat[0];
        mat3d->data[4] = mat[1];
        mat3d->data[12] = mat[4];
        mat3d->data[1] = mat[2];
        mat3d->data[5] = mat[3];
        mat3d->data[13] = mat[5];
        mat3d->data[8] = 0.;
        mat3d->data[9] = 0.;
        mat3d->data[10] = 1.;
        mat3d->data[2] = 0.;
        mat3d->data[6] = 0.;
        mat3d->data[14] = 0.;
        mat3d->data[15] = 1.;
        imodTransModel3D(model, mat3d, NULL, newCen, zscale, doflip);
        return(0);
      }
      k++;
    }

    if (oneLine >= 0) {
      printf("ERROR: End of file or error before line %d of %s\n", oneLine, filename);
      return(-1);
    }

  } else {
    for (k = 0; k < 3; k++) {
      if (fgetline(fin, line, 80) <= 0) {
        printf("ERROR: Reading line %d from  %s\n", k, filename);
        return(-1);
      }
      sscanf(line, "%f %f %f %f", &mat3d->data[k], &mat3d->data[k + 4],
             &mat3d->data[k + 8], &mat3d->data[k + 12]);
      mat3d->data[k + 12] *= transScale;
    }
    mat3d->data[12] += transx;
    mat3d->data[13] += transy;
    mat3d->data[14] += transz;
    imodTransModel3D(model, mat3d, NULL, newCen, zscale, doflip);
  }
  return(0);
}

/*
 * Translate one slice of the model
 *  model  is the model 
 *  mat    is a 6-element matrix with transformation
 *  slice  is the slice number
 *  newCen has the new X and Y center coordinates
 */
static int trans_model_slice(Imod *model, float *mat, int slice, Ipoint newCen)
{
  int ob, co, pt;
  Iobj *obj;
  Icont *cont;
  int zval;
  float x, y;

  float xcen = model->xmax * 0.5f;
  float ycen = model->ymax * 0.5f;

  for (ob = 0; ob < model->objsize; ob++){
    obj = &(model->obj[ob]);
    for (co = 0; co < obj->contsize; co++){
      cont = &(obj->cont[co]);
      for (pt = 0; pt < cont->psize; pt++){
        zval = (cont->pts[pt].z + 0.5f);
        if (zval == slice){
          x = cont->pts[pt].x;
          y = cont->pts[pt].y;
          cont->pts[pt].x = 
            mat[0] * (x - xcen) +
            mat[1] * (y - ycen) +
            mat[4] + newCen.x;
          cont->pts[pt].y =
            mat[2] * (x - xcen) +
            mat[3] * (y - ycen) +
            mat[5] + newCen.y;
        }
      }
    }
  }
  return(0);
}

